### PR TITLE
Respect udp_answer_limit parameter even if UDP payload size is not 512

### LINKED ISSUE #3355: dns.config.udp_answer_limit not taken into account if edns is used
When filing a bug, please include the following:

### `consul version` for both Client and Server
Client: `0.8.5`
Server: `0.8.4`

### `consul info` for both Client and Server
Client:
```
agent:
	check_monitors = 0
	check_ttls = 0
	checks = 1
	services = 2
build:
	prerelease = 
	revision = 2c77151
	version = 0.8.5
consul:
	known_servers = 3
	server = false
runtime:
	arch = amd64
	cpu_count = 32
	goroutines = 41
	max_procs = 2
	os = linux
	version = go1.8.3
serf_lan:
	coordinate_resets = 0
	encrypted = false
	event_queue = 0
	event_time = 3006
	failed = 10
	health_score = 0
	intent_queue = 0
	left = 0
	member_time = 1073708
	members = 3625
	query_queue = 0
	query_time = 1935
```

Server:
```
agent:
	check_monitors = 0
	check_ttls = 0
	checks = 1
	services = 3
build:
	prerelease = 
	revision = f436077
	version = 0.8.4
consul:
	bootstrap = false
	known_datacenters = 8
	leader = false
	leader_addr = 10.71.0.171:8300
	server = true
raft:
	applied_index = 119308751
	commit_index = 119308751
	fsm_pending = 0
	last_contact = 18.798146ms
	last_log_index = 119308751
	last_log_term = 462751
	last_snapshot_index = 119303064
	last_snapshot_term = 462751
	latest_configuration = [{Suffrage:Voter ID:10.71.0.173:8300 Address:10.71.0.173:8300} {Suffrage:Voter ID:10.71.0.171:8300 Address:10.71.0.171:8300} {Suffrage:Voter ID:10.71.0.187:8300 Address:10.71.0.187:8300}]
	latest_configuration_index = 1
	num_peers = 2
	protocol_version = 2
	protocol_version_max = 3
	protocol_version_min = 0
	snapshot_version_max = 1
	snapshot_version_min = 0
	state = Follower
	term = 462751
runtime:
	arch = amd64
	cpu_count = 24
	goroutines = 14691
	max_procs = 23
	os = linux
	version = go1.8.3
serf_lan:
	coordinate_resets = 0
	encrypted = false
	event_queue = 0
	event_time = 3006
	failed = 9
	health_score = 0
	intent_queue = 0
	left = 0
	member_time = 1073708
	members = 3624
	query_queue = 0
	query_time = 1935
serf_wan:
	coordinate_resets = 0
	encrypted = false
	event_queue = 0
	event_time = 1
	failed = 0
	health_score = 0
	intent_queue = 0
	left = 0
	member_time = 66258
	members = 24
	query_queue = 0
	query_time = 1
```

### Operating system and Environment details

centos 7.2 or centos 6.8

### Description of the Issue (and unexpected/desired result)

Consul agent is configured with `dns.config.udp_answer_limit` to 1.
We configure bind to query consul dns api:
```
zone "consul" IN {
	type forward;
	forwarders {
	127.0.0.1 port 8600;
	};
	forward only;
};
```
When asking bind about consul zone, we get more than 1 results.

### Reproduction steps

`dig consul.service.consul @localhost -p 8600` gives 1 result
`dig consul.service.consul @localhost` gives 3 results.

### Early investiguations

After analyzing a tcpdump, it seems that bind is using edns fields to allow udp payload size to be larger than the default 512.
In this case consul seems to ignore udp_answer_limit parameter and return as many results as possible.

This is confirmed by https://github.com/hashicorp/consul/blob/496b0bcf07ad2a6610b1fa3f2db3de00ef9c4eec/agent/dns.go#L570

### PR DESCRIPTION
**Pull request by kamaradclimber**
Fix #3355

Signed-off-by: Grégoire Seux <g.seux@criteo.com>

### CODE REVIEW

**Comment by magiconair** [comment]
Can you add a test for this, please?

**Comment by magiconair** [comment]
@sethvargo this was introduced in a8f1e7ebf. You might want to have a look.

File: agent/dns.go [hunk 1]
```diff
@@ -567,7 +567,7 @@ func trimUDPResponse(config *DNSConfig, req, resp *dns.Msg) (trimmed bool) {
 
 	// This cuts UDP responses to a useful but limited number of responses.
 	maxAnswers := lib.MinInt(maxUDPAnswerLimit, config.UDPAnswerLimit)
-	if maxSize == defaultMaxUDPSize && numAnswers > maxAnswers {
+	if numAnswers > maxAnswers {
```
Comment (sethvargo) [inline agent/dns.go]: I think the idea here is that, if request is edns and the size has changed from the default max size, we don't want to trim the answers or else they will be truncated.

**Comment by kamaradclimber** [comment]
Thanks for the feedback
Ill

Le jeu. 10 août 2017 à 07:59, Seth Vargo <notifications@github.com> a
écrit :

File: agent/dns.go (hunk 1 above)
Comment (kamaradclimber) [reply agent/dns.go]: @sethvargo  I'm not sure to understand what you mean. Would you have an example?

Comment (miraclejianc) [reply agent/dns.go]: I think you can try "+bufsize=512" with dig , because the dig may not build with edns.

**Comment by kamaradclimber** [comment]
abandoning this PR, I'll solve our issue with a different mechanism.

**Pull request closed without merging**